server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/ContinuousComputationTests.java

```diff
@@ -45,47 +46,39 @@ public static void terminateThreadPool() {
         final var computation = new ContinuousComputation<Integer>(threadPool.generic()) {
 
             public final Semaphore executePermit = new Semaphore(1);
+            public final AtomicInteger last = new AtomicInteger();
+            public final AtomicInteger count = new AtomicInteger();
 
             @Override
             protected void processInput(Integer input) {
-                assertTrue(executePermit.tryAcquire(1));
-                result.set(input);
+                assertThat("Should execute computations sequentially", executePermit.tryAcquire(1), equalTo(true));
+                assertThat("Should not reorder executions", input, greaterThan(last.getAndSet(input)));
+                count.incrementAndGet();
                 executePermit.release();
             }
         };
 
-        final Thread[] threads = new Thread[between(1, 5)];
-        final int[] valuePerThread = new int[threads.length];
-        final CountDownLatch startLatch = new CountDownLatch(1);
-        for (int i = 0; i < threads.length; i++) {
-            final int threadIndex = i;
-            valuePerThread[threadIndex] = randomInt();
-            threads[threadIndex] = new Thread(() -> {
-                try {
-                    assertTrue(startLatch.await(10, TimeUnit.SECONDS));
-                } catch (Exception e) {
-                    throw new AssertionError(e);
-                }
-                for (int j = 1000; j >= 0; j--) {
-                    computation.onNewInput(valuePerThread[threadIndex] = valuePerThread[threadIndex] + j);
-                }
-            }, "submit-thread-" + threadIndex);
-            threads[threadIndex].start();
-        }
-
-        startLatch.countDown();
-
-        for (Thread thread : threads) {
-            thread.join();
+        final AtomicInteger listenersComputed = new AtomicInteger();
+        final AtomicInteger inputGenerator = new AtomicInteger(0);
+        final int inputs = scaledRandomIntBetween(10, 10_000);
+
+        for (int i = 0; i < inputs; i++) {
+            final int input = inputGenerator.incrementAndGet();
+            computation.onNewInput(input, () -> {
+                assertThat("Should execute listener after the computation is complete",
+                    computation.last.get(), greaterThanOrEqualTo(input));
+                listenersComputed.incrementAndGet();
+            });
         }
 
         assertBusy(() -> assertFalse(computation.isActive()));
 
-        assertTrue(Arrays.toString(valuePerThread) + " vs " + result.get(), Arrays.stream(valuePerThread).anyMatch(i -> i == result.get()));
+        assertThat("Should keep the latest result", computation.last.get(), equalTo(inputGenerator.get()));
+        assertThat("May skip some computations", computation.count.get(), lessThanOrEqualTo(inputs));
+        assertBusy(() -> assertThat("Should complete all listeners", listenersComputed.get(), equalTo(inputs)));
```

Comment:
Listeners might still be executing after the computation is complete, hence assertBusy here